pkg/llbutils/utils.go

```diff
@@ -123,59 +123,159 @@ func Mkdir(state llb.State, owner string, dirs ...string) llb.State {
+		IgnoreCache:      buildOpts.IgnoreCache,
+		WithCacheMount:   buildOpts.EnableCacheMounts,
+		CacheIDNamespace: buildOpts.CacheIDNamespace,
+	}
+}
+
 // InstallSystemPackages installs the given packages with the given package
 // manager. Packages map have to be a set of package names associated to their
 // respective version.
 func InstallSystemPackages(
 	state llb.State,
 	pkgMgr string,
 	locks map[string]string,
-	ignoreCache bool,
+	opts SystemPackagesCaching,
 ) (llb.State, error) {
 	if len(locks) == 0 {
 		return state, nil
 	}
 
-	var cmds []string
-	var pkgNames []string
-
+	// Packages are sorted by their names first to be sure to always run the
+	// same command for a given set of packages.
+	pkgNames := make([]string, 0, len(locks))
 	for pkgName := range locks {
 		pkgNames = append(pkgNames, pkgName)
 	}
 	sort.Strings(pkgNames)
 
-	packageSpecs := []string{}
+	packageSpecs := make([]string, 0, len(pkgNames))
 	for _, pkgName := range pkgNames {
 		packageSpecs = append(packageSpecs, pkgName+"="+locks[pkgName])
 	}
 
 	switch pkgMgr {
 	case APT:
-		cmds = []string{
-			"apt-get update",
-			fmt.Sprintf("apt-get install -y --no-install-recommends %s", strings.Join(packageSpecs, " ")),
-			"rm -rf /var/lib/apt/lists/*",
-		}
+		return InstallPackagesWithAPT(state, packageSpecs, opts)
 	case APK:
-		cmds = []string{
-			"apk add --no-cache " + strings.Join(packageSpecs, " "),
-		}
+		return InstallPackagesWithAPK(state, packageSpecs, opts)
 	default:
 		return llb.State{}, UnsupportedPackageManager
 	}
+}
+
+func InstallPackagesWithAPT(
+	state llb.State,
+	packageSpecs []string,
+	opts SystemPackagesCaching,
+) (llb.State, error) {
+	runOpts := []llb.RunOption{}
+	cmds := []string{
+		"apt-get update",
+		"apt-get install -y --no-install-recommends " + strings.Join(packageSpecs, " "),
+	}
+
+	if opts.WithCacheMount {
+		cmds = append(cmds, "apt-get autoclean")
+		runOpts = append(runOpts,
+			CacheMountOpt("/var/cache/apt", opts.CacheIDNamespace, "0"),
+			CacheMountOpt("/var/lib/apt", opts.CacheIDNamespace, "0"))
+	} else {
+		cmds = append(cmds, "rm -rf /var/lib/apt/lists/*")
+	}
 
 	stepName := fmt.Sprintf("Install system packages (%s)", strings.Join(packageSpecs, ", "))
-	runOpts := []llb.RunOption{
+	runOpts = append(runOpts,
 		Shell(cmds...),
-		llb.WithCustomName(stepName)}
+		llb.WithCustomName(stepName))
 
-	if ignoreCache {
+	if opts.IgnoreCache {
 		runOpts = append(runOpts, llb.IgnoreCache)
 	}
 
 	return state.Run(runOpts...).Root(), nil
 }
 
+func InstallPackagesWithAPK(
+	state llb.State,
+	packageSpecs []string,
+	opts SystemPackagesCaching,
+) (llb.State, error) {
+	runOpts := []llb.RunOption{}
+	cmds := []string{}
+
+	if opts.WithCacheMount {
+		cmds = append(cmds,
+			"apk add "+strings.Join(packageSpecs, " "),
+			// Clean up old packages but retain installed ones
+			"apk cache -v sync")
+		runOpts = append(runOpts,
+			CacheMountOpt("/etc/apk/cache", opts.CacheIDNamespace, "0"))
```

Comment:
:o apk has cache in `/etc` dir ? unsusual ^^

Reply:
Yep, Alpine wiki states that there's some `setup-apkcache` script available to create a symlink between `/var/cache/apk` and `/etc/apk/cache`. But actually this script doesn't exist on Alpine images, so after some attempts I figured out that only the target directory have to exists to tell apk to cache packages :wink: